### PR TITLE
Improve Credits component

### PR DESCRIPTION
**Pull request by dem4ron**
## Before
Screenshot 2022-10-25 at 18 53 12

## After
Screenshot 2022-10-25 at 18 52 07

Screenshot 2022-10-25 at 18 53 57

Screenshot 2022-10-25 at 18 54 20

Component handles a lot of stuff

if `handle` is missing, falls back to 1 `[top_label]`

Screenshot 2022-10-25 at 18 56 26

### CODE REVIEW

**Review by iHiD — APPROVED** [review]
This looks good to me on the Rails side. I'll let Erik check the JS. Nice work! :)

File: app/javascript/packs/application.tsx [hunk 1]
```diff
@@ -462,6 +461,7 @@ import { highlightAll } from '../utils/highlight'
 import { AutomationLockedTooltipProps } from '../components/tooltips/AutomationLockedTooltip.js'
 import { ApproachesProps } from '@/components/track/Approaches'
 import { ChartData } from '@/components/impact/Chart'
+import { CreditsUsersProp } from '@/components/common/Credits.js'
```
Comment (dem4ron) [inline app/javascript/packs/application.tsx]: just noticing this `.js` at the end..

File: app/helpers/react_components/track/credits.rb [hunk 2]
```diff
@@ -1,16 +1,15 @@
 module ReactComponents
   module Track
     class Credits < ReactComponent
```
Comment (ErikSchierboom) [inline app/helpers/react_components/track/credits.rb]: At some point we should probably rename this to something more generic, as it doesn't necessarily have to do anything with a track. Maybe it should be `Common::Credits`, which is also how it is named in React.

File: app/helpers/react_components/track/credits.rb [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
-      initialize_with :avatar_urls, :top_count, :top_label, :bottom_count, :bottom_label, max: 2, css_class: nil
+      initialize_with :users, :top_count, :top_label, :bottom_count, :bottom_label, css_class: nil
 
       def to_s
         super("track-credits", {
```
Comment (ErikSchierboom) [inline app/helpers/react_components/track/credits.rb]: If you apply the above suggestion, please also change this value.

File: app/javascript/components/common/AvatarGroup.tsx [hunk 4]
```diff
@@ -1,30 +1,30 @@
...
-          <Avatar
-            className="face"
-            src={avatarUrl}
-            key={`${avatarUrl}${Math.random()}`}
-          />
-        ))}
-      {max > 0 && avatarUrls.length - max > 0 ? (
-        <div className="counter">+{Math.min(99, avatarUrls.length - max)}</div>
+    <div className={`c-faces-with-overflow-counter ${className}`}>
+      {users.slice(0, 2).map((user) => (
```
Comment (ErikSchierboom) [inline app/javascript/components/common/AvatarGroup.tsx]: You're guaranteed that users only ever has a maximum of three elements, so the slice is not needed.

File: app/javascript/components/common/AvatarGroup.tsx [hunk 5]
```diff
@@ -1,30 +1,30 @@
 import React from 'react'
 import { Avatar } from '../common'
+import { CreditsUsersProp } from './Credits'
```
Comment (ErikSchierboom) [inline app/javascript/components/common/AvatarGroup.tsx]: This is giving me a TS error that no such type is exported

File: app/javascript/components/common/Credits.tsx [hunk 6]
```diff
@@ -1,38 +1,48 @@
...
-          {pluralizeWithNumber(topCount, topLabel)}
-        </div>
+    <div className={`flex items-center ${className}`}>
+      <AvatarGroup
+        className={`mr-${overflow > 0 ? 12 : 8}`}
+        overflow={overflow}
+        users={users}
+      />
+      <div>
+        {topCount + Number(bottomCount) > 1 || users[0]['handle'].length === 0
```
Comment (ErikSchierboom) [inline app/javascript/components/common/Credits.tsx:38]: `users[0]['handle'].length === 0` I don't think this is ever true. Every user has a handle. Or are you considering the fact that the data might not be there?

File: app/javascript/packs/application.tsx (hunk 1 above)
Comment (ErikSchierboom) [reply app/javascript/packs/application.tsx]: I see a `.js` three lines above this line too

File: app/javascript/components/common/Credits.tsx (hunk 6 above)
Comment (dem4ron) [reply app/javascript/components/common/Credits.tsx:38]: yes, just a fallback in case something would happen.

**Comment by dem4ron** [comment]
Once this will be merged: https://github.com/exercism/website/pull/3116, the flakey test won't fail here.

**Pull request merged**